files/routes/chat.py

```diff
@@ -1,117 +1,271 @@
+typing: list[str] = []
+online: list[str] = []	# right now we maintain this but don't actually use it anywhere
+connected_users = set()
+
+def send_system_reply(text):
+	data = {
+		"id": str(uuid.uuid4()),
+		"avatar": g.db.query(User).filter(User.id == NOTIFICATIONS_ID).one().profile_url,
+		"user_id": NOTIFICATIONS_ID,
+		"username": "System",
+		"text": text,
+		"text_html": sanitize(text),
+		'time': time.time(),
+	}
+	emit('speak', data)
+
+def get_chat_messages():
+	# Query for the last visible chat messages
+	result = (g.db.query(ChatMessage)
+          .join(User, User.id == ChatMessage.author_id)  # Join with the User table to fetch related user data
+          .order_by(ChatMessage.created_datetimez.desc())
+          .limit(CHAT_SCROLLBACK_ITEMS)
+          .all())
+	
+	# Convert the list of ChatMessage objects into a list of dictionaries
+	# Also, most recent at the bottom, not the top.
+	messages = [item.json_speak() for item in result[::-1]]
+	
+	return messages
 
+def get_chat_userlist():
+	# Query for the User.username column for users with chat_authorized == True
+    result = g.db.query(User.username).filter(User.chat_authorized == True).all()
+    
+    # Convert the list of tuples into a flat list of usernames
+    userlist = [item[0] for item in result]
+    
+    return userlist
 
-@app.get('/chat.js')
-@limiter.exempt
-def chatjs():
-	resp = make_response(send_from_directory('assets', 'js/chat.js'))
-	return resp
+@app.get("/chat")
+@is_not_permabanned
+@chat_is_allowed()
+def chat(v):
+	return render_template("chat.html", v=v)
 
 
 @socketio.on('speak')
 @limiter.limit("3/second;10/minute")
-@auth_required
+@chat_is_allowed()
 def speak(data, v):
 	limiter.check()
 	if v.is_banned: return '', 403
 
-	vname = v.username.lower()
-	if vname in muted:
-		if time.time() < muted[vname]: return '', 403
-		else: del muted[vname]
+	text = sanitize_raw(
+		data['message'],
+		allow_newlines=True,
+		length_limit=CHAT_LENGTH_LIMIT,
+	)
+	if not text: return '', 400
 
-	global messages, total
-	text = data[:1000].strip()
-	if not text: return '', 403
-	text_html = sanitize(text)
+	command = chat_command_regex.match(text)
+	if command:
+		command_name = command.group(1).lower()
+		command_parameters = command.group(2)
+		
+		if command_name in commands:
+			commands[command_name](command_parameters)
+		else:
+			send_system_reply(f"Unknown command: {command_name}")
 
-	data={
-		"avatar": v.profile_url,
-		"username": v.username,
-		"namecolor": v.namecolor,
-		"text": text,
-		"text_html": text_html,
-		"text_censored": text,
-		"time": int(time.time())
-	}
-	
-	if v.shadowbanned:
-		emit('speak', data)
-	else:
-		emit('speak', data, broadcast=True)
-		messages.append(data)
-		messages = messages[-50:]
+		return
 
-	total += 1
+	text_html = sanitize(text)
+	quotes = data['quotes']
+	
+	chat_message = ChatMessage()
+	chat_message.author_id = v.id
+	chat_message.quote_id = quotes
```

Comment:
How bad will be the crash if the foreign key constraint gets violated here when we try to commit?

Reply:
Honestly, not at all, it'll just abort and throw away the message. I'm sorta tempted to leave it this way to avoid doing an extra DB call just for validation, although it's not like that's expensive.

Any idea if there's a more efficient way to validate this besides doing a g.db.query()?

Reply:
If you can test it and see that it returns a 500 without any hiccups, go ahead with the current solution.